### PR TITLE
Prometheus Metrics

### PR DESCRIPTION
**Pull request by Bslabe123**
Added a Gauge for watching the prefill backlog size. Note port 9090 is conventional for prometheus metrics.

### CODE REVIEW

**Comment by vipannalla** [comment]
Looks good. Can we do a end-2-end run to ensure there no impact on perf? 
+ Zijun, PTAL .

**Review by FanhaiLu1 — COMMENTED** [review]
Could you share the Prometheus Metrics results?

File: jetstream/core/orchestrator.py [hunk 1]
```diff
@@ -442,6 +449,8 @@ def _prefill_thread(self, idx: int):
       my_transfer_backlog = self._transfer_backlogs[idx]
       # The prefill thread can just sleep until it has work to do.
       request = self._prefill_backlog.get(block=True)
+      self._prefill_backlog_size_metric.set(self._prefill_backlog.qsize())
```
Comment (liurupeng) [inline jetstream/core/orchestrator.py:452]: qq, should we record the metrics in to places?

**Comment by liurupeng** [comment]
> Could you share the Prometheus Metrics results?

@Bslabe123 yep, it would be good if we could verify the metric is added in the container logs, thanks!

File: jetstream/core/orchestrator.py (hunk 1 above)
Comment (liurupeng) [reply jetstream/core/orchestrator.py:452]: I would assume this is the real prefill queue size during runtime @FanhaiLu1 right?

File: jetstream/core/orchestrator.py [hunk 2]
```diff
@@ -421,6 +427,7 @@ def place_request_on_prefill_queue(self, request: ActiveRequest):
     """Used to place new requests for prefilling and generation."""
```
[the remark this answers is not in the document]
Comment (JoeZijunZhou) [reply jetstream/core/orchestrator.py:427]: This is adding requests from JetStream's client to JetStream Orchestrator's prefill backlog. It doesn't relate to engines.

File: jetstream/core/orchestrator.py (hunk 1 above)
Comment (JoeZijunZhou) [reply jetstream/core/orchestrator.py:452]: L430 logs the prefill backlog queue size after a request added to the queue; L452 logs the prefill backlog queue size after a request remove/get from the queue.

**Review by JoeZijunZhou — COMMENTED** [review]
LGTM! Would also like to see the E2E validation result of this metric.

File: jetstream/core/orchestrator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (liurupeng) [reply jetstream/core/orchestrator.py:428]: I see, but where is the API used? I didn't find the usage... so I'm not sure whether we should record metrics here

File: jetstream/core/orchestrator.py (hunk 1 above)
Comment (liurupeng) [reply jetstream/core/orchestrator.py:453]: then I would assume we only need to add it in one place right?

Comment (JoeZijunZhou) [reply jetstream/core/orchestrator.py:453]: IIUC, @Bslabe123 is trying to collect the prefill queue size metric as the first step?

File: jetstream/core/orchestrator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JoeZijunZhou) [reply jetstream/core/orchestrator.py:428]: Which API? This is part of the workflow of the JetStream Decode API.

File: jetstream/core/orchestrator.py [hunk 3]
```diff
@@ -242,6 +246,9 @@ def __init__(
     # Stage 1
     # At first, a request is placed here in order to get prefilled.
     self._prefill_backlog = queue.Queue()
+    self._prefill_backlog_size_metric = prometheus_client.Gauge(
```
Comment (JoeZijunZhou) [inline jetstream/core/orchestrator.py:249]: You can get the decode slots size at line L556 and L587. @liurupeng @Bslabe123 It's similar to the prefill backlog metric.

**Comment by Bslabe123** [comment]
> > Could you share the Prometheus Metrics results?
> 
> @Bslabe123 yep, it would be good if we could verify the metric is added in the container logs, thanks!

Able to validate via also deploying a prometheus server to the cluster to scrape the newly emitted metrics, will follow up with an [ai-on-gke](https://github.com/GoogleCloudPlatform/ai-on-gke) PR with validation instructions/demo

**Comment by Bslabe123** [comment]
> LGTM! Would also like to see the E2E validation result of this metric.

Deployed on GKE with this [maxtext setup](https://github.com/GoogleCloudPlatform/ai-on-gke/tree/main/tutorials-and-examples/inference-servers/jetstream/maxtext/single-host-inference) plus a prometheus server, these are metrics after  1000 concurrent `curl`s to the maxtext http endpoint via `seq 100 | xargs -P 100 -n 1 curl --request POST --header "Content-type: application/json" -s localhost:8000/generate --data '{
    "prompt": "What are the top 5 programming languages",
    "max_tokens": 200
}'`
 
Screenshot 2024-05-09 at 10 35 55 AM

**Comment by Bslabe123** [comment]
> Looks good. Can we do a end-2-end run to ensure there no impact on perf?
> 
> * Zijun, PTAL .

Timed the above 100 requests with  `bash`s `time` and got the following
```
real	0m26.914s
user	0m0.940s
sys	0m1.313s
```
Can do more extensive benchmarking if these numbers seem questionable

**Comment by liurupeng** [comment]
lgtm, thanks @Bslabe123 !

**Review by FanhaiLu1 — APPROVED** [review]
Is Prometheus Metrics mandatory?  If the port already been used or something wrong crash  Prometheus server, what would be impact to jet stream sever? In current logic, it would also crash jetstream sever. 

Can you add flag and only enable  Prometheus Metrics sever and collection if the flag is enabled?

**Review by JoeZijunZhou — APPROVED** [review]

**Pull request merged**